### ISSUE TITLE
Linebreaks weirdly in long ES6 template strings with expressions

### ISSUE THREAD

**Issue by Swizec** [prettier/prettier#3368]
**Prettier 1.8.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgIoCuE8OwAOlDlTmABboBO6Y8DOAvKRdTzlOgFs4SHAHIAIhAYBLdFFHdqAXwqKq8AB4wOOAAYBNOJgA0OACTAYtaZgB0dRs1a3+QpaYCeEAjgDu01FQcdFRfdA9cACM4HAAzaAATHAhYnCFbHABJHAY4AAdczARtGAgcLx8QoKsYzGkoXAqcegA3GKg4NrZ6JJqabyYAcxjS-oEBaRhbXQolEGMQCDyYaWhMZFAsGGRYkKKFyKYwAGs4GABlPOZ6weQYBgI4BYSIMB29p5B6ooYYAAUhgJ0O9UPsQAArTAaABCR1OF0EcAAMvU4CCwd4YHkCDAAEzoz5XBg-ZAgSLoSIeVDQeYgAr1GAAdWkCSsyAAHAAGBYFCBFRlMPJ3B6fXIARwI0lyAPQgyBBIWRQmwseipuqDghGIaKQu1BnxgFOZrNoyFxC3u6ACNwAwhBxsCUFBoGiFgQigAVCnrXUfJRKIA)
```sh
# Options (if any):
--print-width=80
```

**Input:**
```jsx
class Quote {
    character = {
        name: 'Dorian'
    }

    text = `Yes, ${this.character.name}, you will always be fond of me. I represent to you all the sins you have never had the courage to commit.`
}
```

**Output:**
```jsx
class Quote {
  character = {
    name: "Dorian"
  };

  text = `Yes, ${
    this.character.name
  }, you will always be fond of me. I represent to you all the sins you have never had the courage to commit.`;
}

```

**Expected behavior:**

I would prefer a long string to _not_ get split into multiple lines, even if it contains expressions. When there are no expressions, Prettier correctly lets a long string go beyond the `print-width` limit.

I can't decide if this is a bug or a personal preference, but it's been creating weird hard to read code in our codebase. We can clean most of it up by simplifying our expressions, but there's only so short we can go.

**Comment by ikatyang** [comment]
Ref: previous discussion https://github.com/prettier/prettier/issues/3315#issuecomment-346897564

**Comment by duailibe** [comment]
As we'll focus the discussion on this issue, I'll copy my comment from the other thread (cc @lipis):

The thing about template literals is without breaking between `{` and `}` we were actually printing:

```js
const str1 = `text text text ${foo.
  bar} text text`;

const str2 = `text text text ${foo[
  bar
]} text text`;
```

which IMO are strictly worse than what we're printing right now. Note the `foo.bar` and `foo[bar]` are a group with "break points", meaning that if they don't fit in one line, they will break. We'd need to special case printing MemberExpressions when inside a template literal so this wouldn't happen.

The other problem with this is when should and shouldn't break in those cases:

```js
foo = `Text text text text text ${foo.bar.baz.spam.eggs}`;
foo = `Text text text text text ${longFooBarBazSpam.longFooBarBazEggs}`;

**Comment by lipis** [comment]
I think those cases we shouldn't brake at all..

```
foo = `Text text text text text ${foo.bar.baz.spam.eggs}`;
foo = `Text text text text text ${longFooBarBazSpam.longFooBarBazEggs}`;
```

and consider that a _single expression_

while this one is not: 

```
foo = `Text text text text text ${foo.bar.baz.spam.eggs + foo.bar.baz.bacon}`;
foo = `Text text text text text ${longFooBarBazSpam.longFooBarBazEggs - longFooBarBazSpam.longFooBarBazBacon}`;
```

**Comment by Swizec** [comment]
Personally I am of the opinion that a template string, regardless of what’s
in there, is a single expression. People putting too much logic in there is
an antipattern and almost always leads to a mess.

Discouraging messes is the job of a linter, not a formatter, if I’m not
mistaken. Maybe a good way to approach this is to have people vote on
whether they perceive complex string templates as one expression or more?

**Comment by lipis** [comment]
I would agree.. maybe it makes sense to not break them at all..

**Comment by duailibe** [comment]
> Discouraging messes is the job of a linter, not a formatter

The job of a formatter is to format the code, whether or not it's a mess, trying its best to output code that's consistent. I don't think we should be in charge of encouraging or discouraging any pattern. 

And prettier used to not format template literals at all, but people complained code wasn't getting formatted. Then it started using `removeLines` to remove breaks which would output `${a.b.c.d.e}` in a single line, but would break other more complex cases, with function calls and what not. I see what we have today as a good compromise but I agree "simple" cases could get better. The whole point is deciding what is simple and have a sane way to format those cases without adding too much complexity to the code.

See #2046, #1893, #1662, #1626, #821, #1183 and there's more :-)

**Comment by jwbay** [comment]
If there's a rule of thumb needed, I'd propose that inserting one existing value into a template string should never cause breaks. If you're doing something that could create a new value (getters aside), or picks between values, it's complex enough to warrant breaks.

Expressions that shouldn't break:
```js
`${foo}` // this is done, which is awesome!
`${foo.bar}`
`${foo[bar]}`
`${foo[bar][baz]}`
```

There's some interesting middle ground that would be nice to have, but breaks with that rule of thumb:
```js
`${foo()}` // all of the above as function calls with no arguments. i.e. just 'get this computed value'
`${foo || bar}` // to inline defaults without paying the 'break cost'. but with member access chains, this could get long. :shrug:
```

**Comment by danwfeeney** [comment]
I'd say a lot of people, including myself, could dig an option that reverts to previous behaviour i.e. rule to disable formatting string literals completely formatStringLiterals: false.Does this a

**Comment by JonathanWolfe** [comment]
I definitely think that template literals should be treated as one holistic expressions and their contents should never be altered.

##### A travesty of formatting
```js
log(
	chalk.bgRed(
		chalk.white(
			`Covered Lines below threshold: ${
				coverageSettings.lines
			}%. Actual: ${coverageSummary.total.lines.pct}%`
		)
	)
);
```

**Comment by j-f1** [comment]
@JonathanWolfe Would you agree that this should not be formatted?

```js
`Covered Lines below threshold: ${
     coverageSettings
.lines

}%. Actual: ${  coverageSummary

.total.

lines
.
pct
  }%`
```

(that’s [valid JS](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEADAwhAbnATnAEwAIAZASyjgGciAjOAGwgHciYALfK9iBgpIgBJgAHShEJEyDlwBDAOZwAynBgwK8qmIB0DCtTFiAvgFJtRAIJgYAV1kMBwqdjwLlNgLYfZuAJ6GobRgIGHttAL1KLUCxAAdrMQlTVDEQABoQCFj1aCpkUB9cFgAFHwQ8lFksCDICdJBZKhhkADN7KjgM2jkwAGtVJVjZMA1kGFwbTpACCDBW9qmKDtwYYrl5b3mGDoyAKyoADwAhHv6YJVkPOHJKLZ3MmxhYx4AmO6mh3GXkEFpZWl8TCg9ViuAoMAA6rUOMgABwABgyoIgHQhcliP1B1DwOHq+AAjjYyPg1gpNkg2tsph0PGQxhNqRoGHAAIo2EJwd4ZUK0KEEGFIF7cuRkSLyTBeWQ-KDQTkZGwdAAq-wqlI6RiMQA))

**Comment by JonathanWolfe** [comment]
@j-f1: Unfortunately, and it pains me to say this to that example, yes; I would leave that alone. 

The copout answer is because nobody would write something like that, but really, I believe the contents of Template Strings/Literals/Expressions to be intentionally formatted by the author when they're written.

An ideal answer might be to leave the expression location alone, but format them as if it is a standalone line.

Maybe an even simpler answer is to no allow line breaks when formatting template expressions.

**Comment by JonathanWolfe** [comment]
@duailibe: I get that expectation. I would say the best thing to do would be to not linebreak the braces of `${  }` as it just makes the the template literal look awkward afterwards

**Comment by duailibe** [comment]
Yeah that's an option.. If we simply remove the linebreaks, that would be formatted as:

```js
const foo = `some text here ${foo
  .bar.baz} continues here`;
```

Which is worse than what we're doing right now. That's just how it prints a `MemberExpression`. So in that case we'd also need to special case if a `MemberExpression` is inside a `TemplateLiteral`, but what would mean for examples like:

```js
const foo = `some text here ${foo.something({
  key: prop
}).bar.baz()}`;
```

Should it remain as is? That's inconsistent with how we print that case in other places:

```js
foo
  .something({
    key: prop
  })
  .bar.baz();
```

Anyway, we're not opposed on improving that, it's just it's a difficult problem and we'd need a consistent suggestion of different examples and how we should print them. Hopefully that gives more context on the decisions on this subject.

**Comment by JonathanWolfe** [comment]
@duailibe: Sure, I've got lots of time right now so I'll make a big list of before/afters.

**Comment by lachieh** [comment]
It's been a month since @JonathanWolfe's suggestion and a few thumbsups on the post. How do we keep it moving? Someone want to do a PR?

**Comment by framerate** [comment]
I just started using prettier and to immediately turn it off because of this! Definitely willing to help out if necessary!

**Comment by lachieh** [comment]
*tumbleweeds and crickets*

**Comment by lydell** [comment]
A PR exploring ways to improve the oh-so-tricky-to-format template literals would certainly be up for discussion :+1:

**Comment by caub** [comment]
I think with template literal, it shouldn't try to respect `printWidth` at all

Other example
https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAzCFMC8mABgBZoA2AhgBQAkwAslTKQHQBOEArlACY1SASgC+AGgbNWnHvxppMAKkwBGAAxrRAUglMW7LrwEUlqjdt0QADlTABLGAE8RQ4gG4QYkNZh30yUCoOLgB3AAUghDRkECoANwg7Pk9YjGRsKgo0OC8AIw5bAGs4GABlG3soAHNkGA5uHJA+CDB0zOyvOyhsjhgwgqqAWyo2rMaAKzQADwAhArBisqpBuAAZLrhRju9uGCtdgCYtxpsOHpjcqlzHCmgUqw4umAB1JNZkAA41LweIbOeClYYg84D04psvBw4ABHbh2KH9KhDEZIDJjLzZQZ2Wr1RpoLpVChwACK3Ag8GOXhgV1efHeSAOVIKdgoBIAwhBBsMYlBoBCQNxsgAVK7RVHtOAiERAA

**Comment by tgroutars** [comment]
Templates strings are still string, and it's helpful to display them as is, hence never ever add a new line. If you need to call a function in a template string, then just declare a variable.
At least having the option to make it behave that way would be great

**Comment by suchipi** [comment]
@tgroutars we will keep simple variables as one-line, but more "complex" things get broken.

Here is the current behavior (all lines exceed the print width):

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEADAKgCwJYGcAEe+ATnAIYA2FAnvhdAOb4BGFZLbH7r3nPXA-kL4je7ACTAoZALZwAvqgDcAHShqMOAkVKUadRoNFGxJs8NP9JZGDGLZmAV3i4AdNLmLV6qJqI7yKlp6KCYLc2NwqMjjSRkIABM4ClcbOwdnODcPBWU1DSx-Al0gg1CIyxjK6or8OMTk1wY4GAA5WTgACgBKL3zfQu1iwP0QsKra6JqpuuB4pJTmmE6VEBzV3ryoEAAaEAgABxhsaFxkUDJiYggAdwAFS4QzlDIANwhsBN2QMlwYZAAZpRcHA9sxiGQwABrFoAZQOkOwoWQdkcoJACQgYEBwPRSJBxBgdwhDBkZBxFBBewAVrgAB4AIQh0LhHQAMki4BSqftnAdnAAmbnohHEAnIEDMMjMaghb4HeywADqnxgmGQAA4AAx7BUQEFKiEHCUKrJwYivLl7CGhOAAUSgXyQAE5XW7ndayLbYTBLv8kDqQKQAI6ObCkYlkUnkpBAynokEybAo4hova4JEMChwACKjgg8GFe19zBVCTVyAFxYh2AomYAwhAZGSJVBoFaQI4Qehpc84yD5PIgA)
```sh
--range-end 9999999
--range-start 0
```

**Input:**
```jsx
`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${name}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${attributes.name}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${model.attributes.name}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${model.getName()}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${model.get("name")}`;

```

**Output:**
```jsx
`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${name}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${
  attributes.name
}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${
  model.attributes.name
}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${model.getName()}`;

`This is really long bla bla bla bla bla bla bla bla bla bla bla bla bla bla ${model.get(
  "name"
)}`;

```

**Comment by tgroutars** [comment]
@suchipi That's awesome! :)

**Comment by caub** [comment]
@suchipi do you think a configuration like I proposed is possible? because for those occasional large template strings, I really prefer letting my code editor wrap it if needed

**Comment by lachieh** [comment]
@suchipi That's the same behaviour as when this was posted back in v1.8.2. I'm the same as OP—I pretty much never want template literals to wrap.

I understand the opinionatedness of Prettier, but if the idea is to remove the need to think about the code formatting, then wrapping template literals goes against that. I have to think much harder about this:
```js
const example = `${person.fullName}'s parent is called ${
  person.parent.fullName
} and is of age ${person.parent.age}`;
```

than I do about this:
```js
const example = `${person.fullName}'s parent is called ${person.parent.fullName} and is of age ${person.parent.age}`;
```

To get around this, I would have to define three extra variables, even if they only get used this one time. That's not something that will get optimized away in a minifier/webpack.

```js
const personName = person.fullName;
const parentName = person.parent.fullName;
const parentAge = person.parent.age;
const example = `${personName}'s parent is called ${parentName} and is of age ${parentAge}`;
```

What determines whether there should be an option for preferences like this?

**Comment by suchipi** [comment]
@caub I don't see your proposal, could you send me a link to it?

@lachieh we aren't satisfied with this output either, but template literal strings have been a somewhat difficult thing to get right. At one point in the past we removed all newlines and it broke some things (like styled-components). I'd be open to changing the current formatting rules to be more inclusive for which things should not break; I don't think there's a need for an option here (and I would prefer not having one).

**Comment by suchipi** [comment]
We won't do that; please read https://prettier.io/docs/en/option-philosophy.html for more info.

**Comment by caub** [comment]
Ok thanks I know the options must remain very simple. So then I'd suggest not trying to break template literals on their expressions.
And thanks again for that tool

**Comment by zcmgyu** [comment]
I know my answer isn't your problem but I just want to share for somebody try to find same problem with ESLint.
In my `.eslintrc` file I added  `indent: ['error', 'tab'],`, it make my template literal break new line. 
Just remove it

**Comment by nlenkowski** [comment]
Just ran into this myself today... definitely not the formatting I was looking for!

```javascript
console.log(
  `API: Loaded ${properties.data.data.length} Properties / ${
    propertyTypes.data.data.length
  } Property Types`
);
```

Since formatting template literals is pretty tricky I second @caub opinion. Perhaps they shouldn't be formatted at all if they exceed the printWidth?

**Comment by duailibe** [comment]
@nlenkowski This is not an option as we're not interested in regressing from previous developments. As mentioned before in this very thread, Prettier used to not format the expressions and only started because it was very much requested.

**Comment by opennota** [comment]
@duailibe Then, perhaps, this could be made a config option, like singleQuote or trailingComma.

**Comment by nlenkowski** [comment]
@opennota It looks like a config option has already been taken off the table, [see above](https://github.com/prettier/prettier/issues/3368#issuecomment-379434775).

@duailibe Gotcha, makes sense. Occasional wonky formatting of template literals is not the end of the world anyways!

**Comment by caub** [comment]
> Prettier used to not format the expressions

All expression or templates literal expressions?

formatting them in definitely good, it just would be great to inline the inner expressions in it

**Comment by duailibe** [comment]
@opennota A config option is off the table: we don't like options and this issue can be solved without adding options.

@nlenkowski We do want to improve how they're printed though, see my response below.

> formatting inner expressions in templates literals is definitely good, it just would be great to inline them (no multiline/line-breaks)

@caub I agree. It's not impossible, it's just difficult as the printer will occasionally break the nodes inside the expressions (i.e. a very long member chain will naturally break) and we need a way to print them all differently when inside a template expression.

There are ways to achieve that: we could add some `if`s to check if it's inside a template expression but we'd have to add a lot of those, make sure we covered all possible nodes and new changes to the formatter would also have to take into account how they would print if inside a template expression. 

My point is, we just haven't found a solution for this problem yet.

**Comment by morris** [comment]
How about increasing the printWidth while the printer is inside a template string (e.g. printWidth + 40 or some other constant)? This way, really long template strings will still get breaks (which makes sense IMO) but shorter template strings don't break too easily. Not sure if it's easier to implement compared to what @duailibe described but it could strike a nice balance between readability and too long lines. Also 80 + 40 = 120 looks like a nice default for something like this.

**Comment by jacobgasser** [comment]
Same here, Prettier, even though i love you, I have to disable you. I kept trying to do this, 
```toEmail = `<body><h1>Name: ${data.srcElement[0].value}</h1>\n<h1>E-mail / Phone Number: ${data.srcElement[1].value}</h1>\n<h1>Message: </h1>\n${data.srcElement[2].value}</body>`;```


And Prettier kept turning it into this, 
```toEmail = `<body><h1>Name: ${
data.srcElement[0].value
}</h1>\n<h1>E-mail / Phone Number: ${
          data.srcElement[1].value
        }</h1>\n<h1>Message: </h1>\n${data.srcElement[2].value}</body>`;```

Then I would get `SyntaxError: literal not terminated before end of script`. So bye bye for now Prettier, hopefully there is a fix for this soon!

**Comment by j-f1** [comment]
Can you reproduce this in [our playground](https://prettier.io/playground) @jacobgasser? Prettier shouldn’t be causing syntax errors.

**Comment by Marcholio** [comment]
This is the current behavior **(v1.16.1)**

**Input:**
```
toEmail = `<body><h1>Name: ${data.srcElement[0].value}</h1>\n<h1>E-mail / Phone Number: ${data.srcElement[1].value}</h1>\n<h1>Message: </h1>\n${data.srcElement[2].value}</body>`;
```

**Output:**
```
toEmail = `<body><h1>Name: ${
  data.srcElement[0].value
}</h1>\n<h1>E-mail / Phone Number: ${
  data.srcElement[1].value
}</h1>\n<h1>Message: </h1>\n${data.srcElement[2].value}</body>`;
```

So seems like only indentation has changed a bit.

**Comment by lachieh** [comment]
@jacobgasser Your SyntaxError is not a result of Prettier. Something is preventing the browser from loading the full script. The code you pasted is definitely terminated correctly so perhaps some other step in your build process is cutting the script short or the browser just isn't loading the full thing. Either way, it's not a fault of Prettier.

**Comment by brikis98** [comment]
Hey everyone. Prettier is a great tool, but the template literal formatting issue has just gone from "mild annoyance" to total blocker for me. I've created a DSL around [tagged template literals](https://basarat.gitbooks.io/typescript/docs/template-strings.html#tagged-templates) (the code below is simplified for readability):

```typescript
routes: [
  route(parse `GET /widgets`),
  route(parse `GET /widgets/${int("id")}`),
  route(parse `GET /widgets/search/from/${date("from")}/to/${date("to")}?${string("query")}`),
  route(parse `POST /widgets`),
  route(parse `PUT /widgets/${int("id")})
]
```

If those `route` entries get long enough, prettier starts to wrap them, and the result is unreadable (again, the code below is simplified to get the point across):

```typescript
routes: [
  route(parse `GET /widgets`),
  route(parse `GET /widgets/${int(
      "id"
  )}`),
  route(parse `GET /widgets/search/from/${date(
      "from"
  )}/to/${date(
      "to"
  )}?${string("query")}`),
  route(parse `POST /widgets`),
  route(parse `PUT /widgets/${int(
    "id"
  )})
]
```

Good luck making any sense of that!

I tried to go through this issue and related ones and it seems like the stance of the prettier team is:

1. Many users requested template strings to be formatted in the past, so it was added to prettier, and now cannot be removed under.
1. Moreover, adding an option to enable/disable the formatting is off the table, as prettier has a philosophy against exposing too many options.
1. Therefore, we're now waiting on someone to come up with a magical heuristic that somehow intelligently formats template strings when it should and ignores them when it shouldn't.

Is that an accurate summary? If so, does anyone even have the start of an idea for how (3) would work given the huge variety of use cases with template literals, and especially tagged template literals, which are likely to become more and more common? If not, then would the team be open to revisiting either (1) or (2)? 

On a slightly unrelated note, I'll also mention that there doesn't seem to be a way [to disable prettier on blocks of JS/TS code](https://github.com/prettier/prettier/issues/5287), so right now, my only options are (a) put `// prettier-ignore` before every single `route` or (b) don't use prettier.

**Comment by batjko** [comment]
> ```ts
> route(parse `GET /widgets/search/from/${date("from")}/to/${date("to")}?${string("query")}`),
> ```

To be fair, a string like this is already rather bad in terms of readability. Prettier's template logic certainly makes it worse, but this is not a good starting point either.

**Comment by jwbay** [comment]
I opened #5979 to introduce the concept of a "simple" template literal that prints with infinite width. The heuristic just covers identifiers and member access expressions to start.

**Comment by dyst5422** [comment]
Can we just have an option to NOT format template strings?

I think that is a simple solution to something that a lot of people have very different opinions about and is very difficult to put together a good heuristic around

**Comment by duailibe** [comment]
People who commented here, can you try out the playground with @jwbay's changes and report what are the cases Prettier still doesn't format correctly?

Thank you in advance for the cooperation

**Comment by snebjorn** [comment]
It's a lot better. However it still does this.

**Prettier pr-5979**
[Playground link](https://deploy-preview-5979--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAJjgM4AOANgIYCeAYlABQkwBOA-AJRL4us5QA5vgA++TITioBcQvmAAdKPhX4cqfMzad5S1atZwY6VsoAGAEmB8AdKggQARlVYBfAB74rt52-wB6L2s2G2cAL3x2XhDw-B4FEHYE13wAWlSg2ygqAFs4VzMAbj1VVxKVQ2NTcShJaShZYuV8VxAAGhAIMlxoEmRQF1YIAHcABRcEPpQqADcIHEJ2kEdWKjAAayMAZTI1gUFkNnQ4DoALGByKAHVTnHhyNbgtybucGbuaZHASPo6BEjgrBgo1WghyVGQqCoFABHQAViR3AAhVYbba5OAAGRkkOhsJACPcW32FDgAEV0BB4LiYScQLtWADWF9nI4aBRoEsyPxYFcFjBTsgABwABg63IgAKuqzIX25cCZMzgSwAjpT4CCulMQFQSKkGrJZEtDGqcIYQVQwRCkFDaR0ATkcIdWMd7STyerlTa8XSYFRHHzCALkAAmDpsKg4Cj7ADCEBy4K+UGgyo66ABABV-VNbQDXK4gA)
```sh
--parser babylon
```

**Input:**
```ts
function displayFn(str?): string | undefined {
    if (str) {
      return `${str.foobar}x ${str.bar} / ${str.baz ? str.baz : "?"} -- ${str.name}`;
    }
    return undefined;
  }
```

**Output:**
```ts
function displayFn(str?): string | undefined {
  if (str) {
    return `${str.foobar}x ${str.bar} / ${str.baz ? str.baz : "?"} -- ${
      str.name
    }`;
  }
  return undefined;
}
```

**Comment by lachieh** [comment]
#5979 has been merged! 🙌

There's more discussion to be had about what constitutes a 'simple' template string, but this will solve 80% of the situations outlined in @JonathanWolfe's original [proposed solution](https://github.com/prettier/prettier/issues/3368#issuecomment-357312678)

**Comment by sgehrman** [comment]
We need an option to turn off line breaks on template strings.  I'm seeing corruption and eslint complains if it expects trailing commas.

  "trailingComma": "all",

```
    return `${oreIdUrl}/sign#app_access_token=${appAccessToken}&account=${account}&broadcast=${broadcast}&callback_url=${encodeURIComponent(callbackUrl)}&chain_account=${chainAccount}&chain_network=${encodeURIComponent(chainNetwork)}&transaction=${encodedTransaction}${optionalParams}`;
```
is being corrupted with a comma in the string:

```
    return `${oreIdUrl}/sign#app_access_token=${appAccessToken}&account=${account}&broadcast=${broadcast}&callback_url=${encodeURIComponent(
      callbackUrl,    <<===== CORRUPTED WITH A COMMA
)}&chain_account=${chainAccount}&chain_network=${encodeURIComponent(chainNetwork)}&transaction=${encodedTransaction}${optionalParams}`;
```

If I remove the comma, eslint comma-dangle complains.

I want my string templates untouched and not corrupted.  I can't use Prettier now.

**Comment by jwbay** [comment]
@sgehrman It's recommended in documentation to disable all lint rules with which Prettier might conflict: https://prettier.io/docs/en/integrating-with-linters.html Prettier has its own option around trailing commas.

**Comment by sgehrman** [comment]
my eslint and prettier are synced, so it's not a problem. The point is if the line gets broken like that with the comma, and then I fix the comma, now eslint gets confused by this line break and thinks it needs a comma.  So I had to turn of trailing commas all to get it to work at all without having a corrupted string.

**Comment by sgehrman** [comment]
Also I don't understand why line breaking a template string is even considered a good idea.

`hello
     there`

and `hello there` 

are completely different strings.  The first one has a newline char and spaces in there now.  This wrapping completely changes the string.

**Comment by j-f1** [comment]
@sgehrman Your code isn’t being broken, it’s just that allowing a trailing comma in function calls is a relatively new JS feature, so your target environment may not support it.



> Also I don't understand why line breaking a template string is even considered a good idea.

That’s true for the *content* of the string, like `foo` in `` `foo ${bar}` ``; however, the formatting of *interpolations* like `bar` in the example doesn’t affect the value of the string.

**Comment by hermann-punk** [comment]
I started using prettier, liked it a lot, and then found all my template expressions broken up into pieces, making them unreadable. There really needs to be an option to disable this.

**Comment by hermann-punk** [comment]
Apologies! It only applied in a narrow set of expressions and I fixed it
for myself by setting line width to 120 characters, which seems to produce
more comfortable results for me.

**Comment by Domvel** [comment]
Please read the issue https://github.com/prettier/prettier/issues/6192 and https://github.com/prettier/prettier/issues/6262 for a decision.

Take a look to this image: 
![image](https://user-images.githubusercontent.com/35452565/68323223-4d14fd80-00c5-11ea-95e3-d172927a5ead.png)

[Demo](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMAWBLAzgOgEYA2AhoUfsQBQA6UABLQAam3Ok4vEdFK0AkwNevXTZyJYmNJspEmUQoBKQbQC+XZv2XDMuOZOLSDRsnMXKV7OV2vMADAA9NdIXrYwIAZRgAnDFADmFACMAGxKzmpWDk5C6rLGOO5evgHBYeYMNAoA3DQgADQgEAAOMBjQWMigRN7eEADuAAo1CJUoRABuEBgAJgUgeN5EYADWcDAexcN+-sg+AK5whWgwALYEAOqY8FhTYHAerRhlHccAnsjgWJWFflhw3jCNQ-6rRMgAZkQE94UAVlh7AAhIajcYeIirOAAGT8cE+31+IAB9g8MwIcAAivMIPAET8liApt57t5LngSGcCNB+sUUjANr10MgABy2Qp0iD3DZDYqXOlwUkdeGFACOOPgzxKbRARCwAFooHA4D0Vf1vHBxRgNc8iK93kgvgTCvdVhg5t5Fib0ViJfDDYjCTASIyesykAAmQo+IgYAgzADCEFWb0uUGgIpA83uABUSG0jfcVCogA)

It's about separating code and text. And not crossing the line limit. Vertical scrolling during development and reviewing should be avoided. 
Prettier should format this string like other nodes. For example arrays. If an array fits the line width it is formatted in a single line. When you add a new item which will overflow the line width, it will be formatted in multilines by wrapping each item. This is prettier in my opinion.

But please take a look to the other threads / issue (see links above in this post), before you reject this proposal.

**Comment by evilebottnawi** [comment]
@Domvel it is open source so we are welcome to PRs

**Comment by Domvel** [comment]
@evilebottnawi Of course. I would do it with pleasure. But this is just yet in the discussion state. And I don't like rejected pull-requests. :)

**Comment by Domvel** [comment]
@evilebottnawi Of course. I would do it with pleasure. But this is just yet in the discussion state. And I don't like rejected pull-requests. :) (Waste of work)

And btw. My wanted format behavior was already implemented before. It was removed. We should just revert it or add an option. But wait, I don't checked it if the previous behavior is exaclty that what I want. (see #6262)

**Comment by shamilovtim** [comment]
I cannot understand why someone would want a template literal such that

```
`hello my name is ${
myObject.name
}  nice to meet you`
```

this is unambiguously ridiculous. i am comfortable saying that 99% of devs would never want this in their codebase. set your word wrap on your text editor and call it a day.

**Comment by Domvel** [comment]
@shamilovtim Nope, you just don't understand it. The string you posted will be formatted by my proposal like:

```ts
//                                                                             |
`hello my name is ${myObject.name} nice to meet you`;
//                                                                             |
```
(unless your double space, I corrected.)

**Comment by Domvel** [comment]
@shamilovtim Nope, you just don't understand it. The string you posted will be formatted by my proposal like:

```ts
//                                                                             |
`hello my name is ${myObject.name} nice to meet you`;
//                                                                             |
```
(unless your double space, I corrected.)

**Comment by shamilovtim** [comment]
I understand full well that you want it to happen when it hits the column width but that makes no difference, It's still the same code. If it's bad code at 50 characters, it's bad code at 100 characters. There is no value judgment in my mind that suddenly makes this good just because my line is long.

**Comment by Domvel** [comment]
@shamilovtim And you want a ugly wrapped or cutted string. You like vertical scrolling and breaking strings / lines at inappropriate places. Or what's your desired format? Any examples? 
Avoid overlapping the print width / line length. (usually 80 or 120). Is this i'm really ridiculous? Show me your source code please. ...

**Comment by shamilovtim** [comment]
to back up my point that if it's bad code at 50 characters, it's bad code at 100 characters. line 7 is a 128 long column all on one line with prettier.

```javascript
const validationSchema = Yup.object().shape({
    reduceFee: Yup.number()
      .typeError('The fee must be a number')
      .required(`If you don't want to reduce the fee please set the visit fee back to ${originalFee}`)
      .max(
        originalFee,
        `The fee can only be reduced below the original price of ${originalFee}. Please set the fee to ${originalFee} or lower`
      )
  });
```

Under your proposal I would get the following result:
```javascript
const validationSchema = Yup.object().shape({
    reduceFee: Yup.number()
      .typeError('The fee must be a number')
      .required(`If you don't want to reduce the fee please set the visit fee back to ${originalFee}`)
      .max(
        originalFee,
        `The fee can only be reduced below the original price of ${originalFee}. Please set the fee to ${
          originalFee
           } or lower`
      )
  });
```

or maybe it would break my string somewhere else. either way, it would look like a function parameter and obscure the meaning of my code. you have a setting in your code editor for word wrap if you have a problem with this, which is completely compatible with the way prettier treats this now. i'm not sure why instead you'd prefer your preference to be inserted into my code rather than changing a setting in your editor.

**Comment by Domvel** [comment]
@shamilovtim No, that is not correct. Not my proposal. 

It will look like this. (Example print with 80 chars).
```ts
const validationSchema = Yup.object().shape({
  reduceFee: Yup.number()
    .typeError('The fee must be a number')
    .required(
      `If you don't want to reduce the fee please set the visit fee back to ${
        originalFee
      }`
    )
    .max(
      originalFee,
      `The fee can only be reduced below the original price of ${
        originalFee
      }. Please set the fee to ${
        originalFee
      } or lower`
    )
});
```
(I hope there are no misstakes. 👀 )

**Comment by shamilovtim** [comment]
yeah, that's just insane to me. it should be clear what's a parameter in my function and what's a string. this compromises my code and it alters the semantic nature of commas and brackets away from every language norm i've ever seen since i learned on Java 10 years ago. hell i'd rather get a horizontal scrollbar that takes 3 seconds to scroll right for 1000 characters than have this kind of behavior from prettier. that's my stance on this.

**Comment by Domvel** [comment]
@shamilovtim Note: That the code highlighter here on Github is really not the best. Especially for literal template string. Try it in VS Code. (I use VS Code with One Dark Pro btw). And you will see the code and string separated clearly.

**Comment by shamilovtim** [comment]
<img width="902" alt="Screen Shot 2019-11-07 at 2 41 05 PM" src="https://user-images.githubusercontent.com/3495974/68421650-ae43dc00-016c-11ea-9e51-e29ac5041121.png">

vs. 

<img width="583" alt="Screen Shot 2019-11-07 at 2 40 49 PM" src="https://user-images.githubusercontent.com/3495974/68421674-b439bd00-016c-11ea-9dff-388dd874823f.png">


using [my theme](https://marketplace.visualstudio.com/items?itemName=timshamilov.dawnchorus) (shameless plug 🙊)

**Comment by Domvel** [comment]
@shamilovtim You ignores the print width. And no one should scroll vertically. I don't even wish that to my enemies. ;)

**Comment by shamilovtim** [comment]
i'm on a 15" 2018 MBP and the 128 column only ends at 50% of my screen.

**Comment by Domvel** [comment]
@shamilovtim Hey, if you set a high value for your print width my proposal will not annoy you. Now again, I am not sure if you have fully understood my proposal. Ok, maybe you want a small print width for the other stuff, but not for the template string, right? That's a bit contradictory. But ok, it's your opinion about this. If you want to fill your full screen, you should write all your code (nodes) in a single line.

**Comment by shamilovtim** [comment]
my print width for prettier is about 120 (50% of my screen) which i think is very reasonable. and the literal i have there on line 7 only breaks it by about 8 characters. i don't mind the param being on a separate line, because both params are being treated equally. it's only once you start treating string literals like params that i have a problem because it's competing with my params for visual meaning. if my string wrapped i'd say so be it. the whole thing would be green in my screenshot, i wouldn't be seeing "originalFee" wrapped down in its lexical color two times like it's a parameter. in my color theme i treat literal syntax highlighting like Monokai does and i'm sure you understand how popular Monokai is with people. this would obfuscate code for thousands of people. if i had to change the syntax highlighting of literals just for prettier i'd say prettier is doing something wrong.

sorry for all the chatter on here.

**Comment by intervalia** [comment]
Why not just allow us to use `/ prettier-ignore-start` and `/ prettier-ignore-end`.

I find ANY changes to my template literals to be an offense against me, God and Nature. :)

I should be allowed to disable the formatting ANY time in my code by wrapping it. But even in version 1.16.4 it refuses to accept the ignore command in my JS.

PLEASE allow this to work!!

**Comment by thorn0** [comment]
@intervalia `// prettier-ignore` [does work](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgIoCuE8OwAOlDjmABboBO6Y89OAvKRVVVOgLZwkOAOQARCPQCW6KMK44AvhXkB6FTgAO9ODBiS49ALSSA5lAlx58AB4x2OAAYBNOJgA0OACTlK3KjBpJTAA6eT9w2gYmFlDfcPiEnF4BeQUPAE8IAhwAd0lUVBx0VBz0dNwAIzgcADNoABMcCBqcAWCcAEkcbS1XBDsYCBxM7OLCgOrMSShcEZw6ADdqqDgl1jpGieosxhNqwe2+PkkYYIcKBRA3EAgNPWhMZFAGeggcgAUGBEeUdAWIST1K4gCqMMAAax0AGUNExpiZkDB6AQ4NcaDA+KgAOqBeCYWFgOBQ74nSQLE7pZDgbDA6aYAwwd67PjoZA1Yr064AK0w1gAQmDITAofw4AAZaZwNkc1EgHnWKHw1BwQjEKVIdmoTkgWH0en0KkVdBVVDArTTGBYwEBZAADgADNctBB6VjGBoqb19UtgQBHIjwJm3H4gLCGFZweqR4Haf2SbRM9AmFnSrWy+nHRHI9NKlUB9Wa7UwY1W+o2pAAJmuSPQ+XhAGEIEdWShzCtgQR6QAVY0-QuyhYojpQKOwKFgKR3ACCI6hMHSytT9IUCiAA) for template literals. Not sure what you need `prettier-ignore-start` and `prettier-ignore-end` for. Could you elaborate? Preferably in a separate issue.

**Comment by intervalia** [comment]
I am using VSCode. I have Prettier 1.16.4.

If I have this in my file:

```
  // prettier-ignore
  return `
  <a class="${mobileDisplay === "false" ? styles.mobileHide : ""}" href="${item.link && item.link.URL ? item.link.URL : ""}" ${item.link && item.link.URL.indexOf("//") > -1 ? `target="_blank"` : ""}>
    <div class="${styles.icon}">${icon({ pkg: "std", name: item.name, title: item.svgTitle})}</div>
  </a>`;
```

And format the file, or even paste into that template literal then the code reformats to this:

```
  // prettier-ignore
  return `
  <a class="${mobileDisplay === "false" ? styles.mobileHide : ""}" href="${
    item.link && item.link.URL ? item.link.URL : ""
  }" ${item.link && item.link.URL.indexOf("//") > -1 ? `target="_blank"` : ""}>
    <div class="${styles.icon}">${icon({
    pkg: "std",
    name: item.name,
    title: item.svgTitle
  })}</div>
  </a>`;
}
```

**Comment by thorn0** [comment]
@intervalia Just tested this with Prettier 1.16.4 and it worked for me. Anyways, it's really off topic in this issue. Please open a separate issue and provide clear steps to reproduce the problem if you want help with this.

**Comment by dawnmist** [comment]
I think that printWidth should still be respected for "simple" cases in some circumstances.

We have one template string that is providing an error message with multiple fields that represent a data key, and were using typescript/tslint with 'prefer template strings'. Previously the template string was separated onto 5 lines, and fairly readable. With prettier 1.19.1 it's condensed into a single line that is 316 characters long - with a print width of 120 characters. I can understand the argument for exceeding the print width by a moderate amount in order to keep the template string together, but almost tripling the print width is arguably taking that too far.

Can I suggest at least putting in a heuristic where if the printWidth is exceeded by too much - e.g. more than 50% - then even the simple template strings should be wrapped?

**Comment by thorn0** [comment]
@dawnmist We might consider something like that, but does it really make sense to use template literals in such situations? Wouldn't a simple concatenation be more readable?

**Comment by dawnmist** [comment]
@thorn0 The tslint rule for prefer template strings marks plain concatenation as undesirable/needs fixing. Until now that rule had worked hand-in-hand nicely with prettier.

It's only the change in prettier to ignore the printWidth entirely for template strings that now makes it a problem - the template strings _used_ to be readable but now are being formatted in a manner that makes no sense given the content of the template string. Sure, changing them all to simple concatenation _now_ would be more readable than what prettier is creating - but that's because prettier is ignoring its other settings entirely and mashing everything into a single line regardless of what the final length would be even in situations when doing that is clearly unsuitable.

**Comment by doplumi** [comment]
most of the time I log I use formatter functions on the content 

```console.log(`... ${format(x)} ...`)```

which ends up as 

```
console.log(`...${
  format(x)
}...`);
```

undesirable in my case. I'd rather have template literals left untouched. It should probably be an option:

```{
  singleQuotes: true,
  ...
  ignoreTemplateLiterals: true,
}```

**Comment by nicu-chiciuc** [comment]
@doplumi If you have a line length of `35` there's definitely no problem. The problem we're facing and I think this issue is about, is when you have a line of 200+ characters which could've been easily split.

**Comment by tgroutars** [comment]
@nicu-chiciuc you see what they meant, I don't think they were saying it's a problem with a line length of 35

**Comment by terrisgit** [comment]
It's difficult to follow all these posts. What is the behavior for the current version of prettier and are there options to override it? In my case I have a print width of 120 and would like reformatting only of lines that exceed this length and contain a legit ${...}

**Comment by j-f1** [comment]
@terrisgit can you open a new issue by pasting some sample code into [the playground](https://prettier.io/playground/), making sure that it demonstrates the undesired behavior you’re seeing, and clicking the “report issue” button at the bottom right? It’s hard to figure out exactly what’s causing the issue unless we can see in detail what’s happening.

**Comment by terrisgit** [comment]
@j-f1 The playground link in the first post is pretty good. It doesn't work though.

```
Error: Invalid proseWrap value. Expected "always", "never" or "preserve", but received true.
    at Ld._applyNormalization (https://prettier.io/lib/standalone.js:1:150240)
    at r (https://prettier.io/lib/standalone.js:1:149535)
    at Ld.normalize (https://prettier.io/lib/standalone.js:1:149575)
    at Yh (https://prettier.io/lib/standalone.js:1:176180)
    at Object.normalizeApiOptions (https://prettier.io/lib/standalone.js:1:177688)
    at normalize (https://prettier.io/lib/standalone.js:1:203877)
    at JD (https://prettier.io/lib/standalone.js:1:227296)
    at Object.formatWithCursor (https://prettier.io/lib/standalone.js:40:40390)
    at formatCode (https://prettier.io/worker.js:141:21)
    at handleMessage (https://prettier.io/worker.js:78:26)
```

**Comment by terrisgit** [comment]
@j-f1 Wow the playground is awesome. Here it is. I think @Swizec got what he wanted (why is this issue still open) and now I don't want it. I can just use // prettier-ignore, that's fine.

[Another example](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgIoCuE8OwAOlDlTmABboBO6Y8DOAvKRdTzlOgFs4SHAHIAIhAYBLdFFHdqAXwqKq8AB4wOOAAYBNOJgA0OACTAYtaZgB0dRs1a3+QiktMBPCARwB3aVRUHHRUP3RPXAAjOBwAM2gAExwIOJwhWxwASRwGOAAHPMwEbRgIHG9fUOCrWMxpKFxKnHoAN1ioOHa2emTamh8mAHNYsoGBAWkYW113EGMQCHyYaWhMZFBGBgg-AAVGBHWUUPDI+ZAopjAAazgYAGVBOAAZBrhkONDihcvmW4f8swGkNkDAGAQ4As4AIYolEnBEs85EMCOgRgAxKQCdAwFZQEHHAhlc60GACVAAdWs8EwgLAcHuhym0laU08yHA2HODWKDBgu2G2I+X0hIAAVpgNPdgag4IRiO8kJ9UN8QICGLyOVF0DFUOdChBihSmPkOYUjHAGO1zgBHIjwAVLI4gLAAWk6CIR5zydukeQFaKFSpFC2Kk1B4NF9Xxsvl8GFKtFMB1FOkiSsyAATAswehAsCAMIQCboDlGACs5wIxQAKjqjsrVa0IVkoPDYPcwDJlgBBNv3GCeWUJ1WFBowVPp2jIABsAAYlEogA)

**Comment by MurzNN** [comment]
With simple template strings seems now works well, but if template contains function - Prettier breaks it to several string, here is modified previous example: [String with template, containing function](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgIoCuE8OwAOlDlTmABboBO6Y8DOAvKRdTzlOgFs4SHAHIAIhAYBLdFFHdqAXwqKq8AB4wOOAAYBNOJgA0OACTAADqgJNU0gF5wAdNdvpUAChi1pmZ3SMzKzO-EIAlEqmAJ4QBDgA7tKoqDgeCejRuABGcDgAZtAAJjgQ+ThCzjgAkjgMcJb1mAjaMBA4sfEeqT55mNJQuJ049ABueVBw42z0Jb00cUwA5nltCwIC0jDOuhRKIMYgEJYw0tCYyKCMDBAJAAqMCBco6ZkXh9lMYADWcDAAyoI4AAZAZwZD5DzND5fX4AyzMAZLZAwBgEOCHOACXJFIpwIrAuRLAjoFYAMSkAnQMFOUGRLwIbQOIFoMAEqAA6r54JgEWA4P8nltpKMttFkOBsMyBs0GDA7ssqRCoRiQAArTAaf5I1BwQjEcFISGoaEgBEMWUS7LoXKoZmNAYwDnSIo+ZAANgADIdGhBmhymJYJY0jHAGONmQBHIjwBXHZ4gLAAWkm+PxzPq0ek9QVpKVRpVh2amxRaNV-Tpuv18GVJtVMBtztdtGQACZDqj0MkkQBhCAbdASowAVmZBGaABUbc9jabRujqlA8bB-mAZCcAIJL-4waK62vNJRKIA)

Is this fixable with current version using some configuration options, or maybe some workaround exists?

**Comment by masnormen** [comment]
Sadly, despite the "great demand" from the community, the PR #12505 isn't approved. This is still one thing from Prettier that bugs me. Template literals are just that: literals.

If a person doesn't intend to add a newline in a template literal, then, maybe, it _should look_ like there's no new line, no? Or maybe at least give us a choice. Forcing to use concatenation instead of template literals for really long string also isn't always readable.

Despite what you said in the Option Philosophy, I just can hope you'd consider reopening that PR 🙏

**Comment by micahbf** [comment]
Just wanted to chime in here with more support for adding this option. 

I am also generally in the camp of less options are better, just give me a reasonable format and don't make me think about it, which is why I love prettier!

But the line splitting on template literals really drives me nuts. It almost always makes the code harder to read and understand. I would really appreciate an option for this behavior!

**Comment by andrew-boyd** [comment]
Found this discussion because I would also like to see this option.

**Comment by syntastical** [comment]
Yeah, put my in the camp of adding configuration for this.  I really dislike the way this is handled.

**Comment by adheeshfordev** [comment]
Just adding my name to the people who expect better handling of template literals.

**Comment by Matthew-Hiebing** [comment]
Adding my name here too.

**Comment by woosuk288** [comment]
OMG, I think many people including me still feel pain from the backtick problem. 

Shouldn't Prettier make code prettier?

**Comment by bakkot** [comment]
Here's a heuristic I haven't seen proposed which I think would make ~everyone happy:

Only add linebreaks to an interpolation if there's already a linebreak within the interpolation.

This is similar to [how](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAKYACAhkjAGAGgwCNcADAHSjIwF8BKAbiqvW10JPJocZAJAQADjACW0AM7JQWAE6yIAdwAKchFJRYANoqwBPKQOKysYANZwYAZSwBbOABlRUOMgBm2iXCMnzlq0KmzgDmyDCyAK7eIF62omGR0RIhWnAAihEQ8O6e0QBWEgAeVinpmdlIHlpeAgCO5XDKCkIaIFgSALQucAAmvfwg4ViiWiEAwhC2tljIbVpaA8lQwakAgjDhosQR8Mpwsk4uOdXRABYwtloA6qei8BKBYHBW6neiAG53erNgEoYg7yiAEkoH1YFYwLJRCJVqCrDA9KljjUQEIFF4riYhLM0XAvLJ3q4BM58TAmlhgtNkdFArJ8bNiFhiHAFgI0c4YFdRD0YKdkAAOQggWRweqiEXkykzSq5AQwJlcnl8pAAJgEES8ABUmRoqii4LZmT0+j0HFhlhEKXAAGIQWTTDYhWZYHYQEC0WhAA) an object expression will get put on multiple lines only if it exceeds print width or contains linebreaks, just with an "and" instead of an "or".

That way if you have a complex interpolation which you want prettier to break up for you, you can put a linebreak in. And if you don't want it broken up - for example, if you wrote the whole template on one line - then prettier will leave it alone.

Thoughts? I'm happy to try a PR if this sounds good.

**Comment by bakkot** [comment]
In [version 3.2.0](https://prettier.io/blog/2024/01/12/3.2.0.html#avoid-introducing-linebreaks-in-template-interpolations-15209httpsgithubcomprettierprettierpull15209-by-bakkothttpsgithubcombakkot), Prettier incorporates the heuristic I suggested in my previous comment: it will no longer add linebreaks to interpolations in template literals unless there is already one present. So if you want to keep the whole string on a single line, just write it that way, and Prettier will respect that. (It will still format the interpolations as normal, just effectively with infinite `printWidth`).

This won't make everyone happy, but I hope it's an improvement for most people in this thread.

**Comment by masnormen** [comment]
> In [version 3.2.0](https://prettier.io/blog/2024/01/12/3.2.0.html#avoid-introducing-linebreaks-in-template-interpolations-15209httpsgithubcomprettierprettierpull15209-by-bakkothttpsgithubcombakkot), Prettier incorporates the heuristic I suggested in my previous comment: it will no longer add linebreaks to interpolations in template literals unless there is already one present. So if you want to keep the whole string on a single line, just write it that way, and Prettier will respect that. (It will still format the interpolations as normal, just effectively with infinite `printWidth`).
> 
> This won't make everyone happy, but I hope it's an improvement for most people in this thread.

That's a simple yet amazing improvement. Thank you for your work

**Comment by Lehoczky-peakfs** [comment]
I wonder if this could be closed since https://github.com/prettier/prettier/pull/15209 has been merged. The template literal formatting is great now.

**Comment by kachkaev** [comment]
@Lehoczky-peakfs, yep, makes sense, thanks! Here is how Prettier 3.3 formats the initial example:

**Prettier 3.3.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgIoCuE8OwAOlDlTmABboBO6Y8DOAvKRdTzlOgFs4SHAHIAIhAYBLdFFHdqAXwqKq8AB4wOOAAYBNOJgA0OACTAYtaZgB0dRs1a3+QpaYCeEAjgDu01FQcdFRfdA9cACM4HAAzaAATHAhYnCFbHABJHAY4AAdczARtGAgcLx8QoKsYzGkoXAqcegA3GKg4NrZ6JJqabyYAcxjS-oEBaRhbXQolEGMQCDyYaWhMZFAsGGRYkKKFyKYwAGs4GABlPOZ6weQYBgI4BYSIMB29p5B6ooYYAAUhgJ0O9UPsQAArTAaABCR1OF0EcAAMvU4CCwd4YHkCDAAEzoz5XBg-ZAgSLoSIeVDQeYgAr1GAAdWkCSsyAAHAAGBYFCBFRlMPJ3B6fXIARwI0lyAPQgyBBIWRQmwseipuqDghGIaKQu1BnxgFOZrNoyFxC3u6ACNwAwhBxsCUFBoGiFgQigAVCnrXUfJRKIA)

```sh
--parser babel
--trailing-comma none
--prose-wrap always
```

**Input:**

```jsx
class Quote {
    character = {
        name: 'Dorian'
    }

    text = `Yes, ${this.character.name}, you will always be fond of me. I represent to you all the sins you have never had the courage to commit.`
}
```

**Output:**

```jsx
class Quote {
  character = {
    name: "Dorian"
  };

  text = `Yes, ${this.character.name}, you will always be fond of me. I represent to you all the sins you have never had the courage to commit.`;
}

```
